File: lib/services/prediction_service.dart
/// سرویس پیش‌بینی آفلاین — میانگین متحرک از SQLite محلی
/// بدون نیاز به اینترنت، همیشه در دسترس

import 'dart:convert';
import 'dart:math' as math;
import 'package:drift/drift.dart';
import '../data/local/database.dart';
import '../data/local/tables/prediction_cache_table.dart';
import '../domain/models/prediction.dart';

const _kDailyLimit = 3;

class PredictionService {
  final AppDatabase _db;

  PredictionService(this._db);

  // ─── محاسبه فروش روزانه از SQLite ─────────────────────────────

  Future<List<Map<String, dynamic>>> getDailySales({int days = 90}) async {
    final since = DateTime.now().subtract(Duration(days: days));
    final invoices = await _db.invoicesDao.getInvoicesByPeriod(since, DateTime.now());

    final Map<String, double> byDay = {};
    for (final inv in invoices) {
      final d = inv.createdAt;
      final key = '${d.year}-${d.month.toString().padLeft(2, '0')}-${d.day.toString().padLeft(2, '0')}';
      byDay[key] = (byDay[key] ?? 0) + inv.finalAmount;
    }

    return byDay.entries
        .map((e) => {'date': e.key, 'amount': e.value})
        .toList()
      ..sort((a, b) => (a['date'] as String).compareTo(b['date'] as String));
  }

  // ─── پیش‌بینی با میانگین متحرک ────────────────────────────────

  Future<SalesForecast> predictSalesOffline({int days = 30}) async {
    final salesData = await getDailySales(days: 90);
    final values = salesData.map((s) => (s['amount'] as num).toDouble()).toList();

    final window = values.isEmpty ? 1 : values.length.clamp(1, 7);
    final recent = values.reversed.take(window).toList();
    final avg = recent.isEmpty ? 0.0 : recent.fold(0.0, (a, b) => a + b) / recent.length;

    double std = 0;
    if (recent.length > 1) {
      final variance = recent.fold(0.0, (a, b) => a + (b - avg) * (b - avg)) / recent.length;
      std = math.sqrt(variance);
    }

    final predictions = <DailyPrediction>[];
    final base = DateTime.now();
    for (int i = 1; i <= days; i++) {
      final d = base.add(Duration(days: i));
      predictions.add(DailyPrediction(
        date: d,
        predictedAmount: avg,
        confidenceLow: (avg - std * 1.5).clamp(0, double.maxFinite),
        confidenceHigh: avg + std * 1.5,
      ));
    }

    final trend = _computeTrend(values);
    final (bestDay, worstDay) = _computeBestWorstDay(salesData);

    return SalesForecast(
      predictions: predictions,
      trend: trend,
      bestDay: bestDay,
      worstDay: worstDay,
      dataSource: 'moving_average',
      dataPoints: salesData.length,
    );
  }

  // ─── هشدار موجودی از SQLite ─────────────────────────────────

  Future<List<StockAlert>> getStockAlertsOffline() async {
    final products = await _db.productsDao.getLowStockProducts();
    if (products.isEmpty) return [];

    final since = DateTime.now().subtract(const Duration(days: 30));
    final invoices = await _db.invoicesDao.getInvoicesByPeriod(since, DateTime.now());

    final Map<int, double> productSales = {};
    for (final inv in invoices) {
      final items = await _db.invoicesDao.getInvoiceItems(inv.id);
      for (final item in items) {
        productSales[item.productId] = (productSales[item.productId] ?? 0) + item.quantity;
      }
    }

    final alerts = <StockAlert>[];
    for (final p in products) {
      final totalSold = productSales[p.id] ?? 0;
      final dailyAvg = totalSold / 30;
      if (dailyAvg <= 0) continue;

      final daysLeft = p.stockQuantity / dailyAvg;
      if (daysLeft > 7) continue;

      final urgency = daysLeft <= 2 ? 'critical' : (daysLeft <= 5 ? 'warning' : 'info');
      alerts.add(StockAlert(
        productId: p.id,
        productName: p.name,
        currentStock: p.stockQuantity.toDouble(),
        dailyAvgSales: dailyAvg,
        daysRemaining: daysLeft,
        urgency: urgency,
      ));
    }
    return alerts..sort((a, b) => a.daysRemaining.compareTo(b.daysRemaining));
  }

  // ─── پرفروش‌ترین محصولات از SQLite ──────────────────────────

  Future<List<TopProduct>> getTopProductsOffline() async {
    final now = DateTime.now();
    final thisWeekStart = now.subtract(const Duration(days: 7));
    final prevWeekStart = now.subtract(const Duration(days: 14));

    Future<Map<int, Map<String, dynamic>>> weekSales(
        DateTime start, DateTime end) async {
      final invs = await _db.invoicesDao.getInvoicesByPeriod(start, end);
      final Map<int, Map<String, dynamic>> sales = {};
      for (final inv in invs) {
        final items = await _db.invoicesDao.getInvoiceItems(inv.id);
        for (final item in items) {
          sales.putIfAbsent(
              item.productId, () => {'name': item.productName, 'qty': 0.0});
          sales[item.productId]!['qty'] =
              (sales[item.productId]!['qty'] as double) + item.quantity;
        }
      }
      return sales;
    }

    final current = await weekSales(thisWeekStart, now);
    final previous = await weekSales(prevWeekStart, thisWeekStart);

    if (current.isEmpty) return [];

    final sorted = current.entries.toList()
      ..sort((a, b) =>
          (b.value['qty'] as double).compareTo(a.value['qty'] as double));

    return sorted.take(5).map((e) {
      final prevQty = (previous[e.key]?['qty'] as double?) ?? 0;
      final currQty = e.value['qty'] as double;
      final trendPct = prevQty > 0 ? (currQty - prevQty) / prevQty * 100 : 100.0;
      final sign = trendPct >= 0 ? '+' : '';
      return TopProduct(
        productId: e.key,
        productName: e.value['name'] as String,
        predictedSales: currQty * 1.05,
        trendPercent: trendPct,
        trendLabel: '$sign${trendPct.toStringAsFixed(0)}٪ نسبت به هفته قبل',
      );
    }).toList();
  }

  // ─── cache AI در SQLite ──────────────────────────────────────

  Future<AiAnalysis?> getCachedAnalysis() async {
    final cached = await _db.predictionDao.getValidCache();
    if (cached == null) return null;
    final recs = (jsonDecode(cached.recommendationsJson) as List).cast<String>();
    return AiAnalysis(
      analysis: cached.analysisText,
      recommendations: recs,
      source: 'cache',
      generatedAt:
          '${cached.createdAt.year}/${cached.createdAt.month}/${cached.createdAt.day}',
      expiresAt:
          '${cached.expiresAt.year}/${cached.expiresAt.month}/${cached.expiresAt.day}',
      remainingRequests: await getRemainingRequests(),
    );
  }

  Future<void> saveAnalysisToCache(AiAnalysis analysis) async {
    final expires = DateTime.now().add(const Duration(hours: 24));
    await _db.predictionDao.saveCache(AiAnalysisCacheTableCompanion(
      analysisText: Value(analysis.analysis),
      recommendationsJson: Value(jsonEncode(analysis.recommendations)),
      salesHash: Value(_dailyKey()),
      createdAt: Value(DateTime.now()),
      expiresAt: Value(expires),
    ));
  }

  // ─── سهمیه ──────────────────────────────────────────────────

  Future<int> getRemainingRequests() async {
    final log = await _db.predictionDao.getTodayUsage();
    final used = log?.requestCount ?? 0;
    return (_kDailyLimit - used).clamp(0, _kDailyLimit);
  }

  Future<void> incrementLocalUsage() => _db.predictionDao.incrementUsage();

  // ─── helpers ────────────────────────────────────────────────

  String _dailyKey() {
    final d = DateTime.now();
    return '${d.year}-${d.month}-${d.day}';
  }

  String _computeTrend(List<double> values) {
    if (values.length < 4) return 'stable';
    final half = values.length ~/ 2;
    final firstAvg = values.take(half).fold(0.0, (a, b) => a + b) / half;
    final secondAvg =
        values.skip(half).fold(0.0, (a, b) => a + b) / (values.length - half);
    if (firstAvg == 0) return 'stable';
    final change = (secondAvg - firstAvg) / firstAvg;
    if (change > 0.05) return 'rising';
    if (change < -0.05) return 'falling';
    return 'stable';
  }

  static const _persianWeekdays = [
    'دوشنبه', 'سه‌شنبه', 'چهارشنبه', 'پنجشنبه', 'جمعه', 'شنبه', 'یکشنبه',
  ];

  (String, String) _computeBestWorstDay(List<Map<String, dynamic>> data) {
    final Map<int, List<double>> byDay = {};
    for (final s in data) {
      final d = DateTime.parse(s['date'] as String);
      final idx = d.weekday % 7;
      byDay.putIfAbsent(idx, () => []);
      byDay[idx]!.add((s['amount'] as num).toDouble());
    }
    if (byDay.isEmpty) return ('—', '—');
    final avgs = byDay.map(
        (k, v) => MapEntry(k, v.fold(0.0, (a, b) => a + b) / v.length));
    final best = avgs.entries.reduce((a, b) => a.value >= b.value ? a : b).key;
    final worst = avgs.entries.reduce((a, b) => a.value <= b.value ? a : b).key;
    return (_persianWeekdays[best], _persianWeekdays[worst]);
  }
}

File: backend/services/prediction_service.py
"""
سرویس پیش‌بینی فروش
ابتدا Prophet امتحان می‌کند، در صورت عدم وجود به moving average برمی‌گردد.
تمام داده‌ها به‌صورت داینامیک از دیتابیس خوانده می‌شوند.
"""
import json, hashlib
from datetime import datetime, timedelta, date
from typing import List, Dict, Any, Optional
from collections import defaultdict
from sqlalchemy.orm import Session
from sqlalchemy import func, cast, Date


def _get_daily_sales(db: Session, days: int = 90) -> List[Dict]:
    """دریافت فروش روزانه از دیتابیس (کاملاً داینامیک)"""
    from models.invoice import Invoice
    since = datetime.utcnow() - timedelta(days=days)

    rows = (
        db.query(
            cast(Invoice.created_at, Date).label("sale_date"),
            func.sum(Invoice.final_amount).label("total"),
        )
        .filter(Invoice.created_at >= since)
        .filter(Invoice.status == "completed")
        .group_by(cast(Invoice.created_at, Date))
        .order_by(cast(Invoice.created_at, Date))
        .all()
    )
    return [{"date": str(r.sale_date), "amount": float(r.total or 0)} for r in rows]


def _moving_average_predict(sales_data: List[Dict], days: int = 30) -> Dict:
    """پیش‌بینی با میانگین متحرک ۷ روزه — کاملاً آفلاین"""
    values = [s["amount"] for s in sales_data]
    if not values:
        values = [0]

    window = min(7, len(values))
    avg = sum(values[-window:]) / window
    std = 0.0
    if window > 1:
        variance = sum((v - avg) ** 2 for v in values[-window:]) / window
        std = variance ** 0.5

    predictions = []
    base = date.today()
    for i in range(1, days + 1):
        d = base + timedelta(days=i)
        predictions.append({
            "date": d.isoformat(),
            "predicted_amount": round(avg, 0),
            "confidence_low": round(max(0, avg - std * 1.5), 0),
            "confidence_high": round(avg + std * 1.5, 0),
        })
    return {"predictions": predictions, "source": "moving_average"}


def predict_sales(db: Session, days: int = 30) -> Dict:
    """پیش‌بینی فروش — Prophet اگر موجود باشد، وگرنه moving average"""
    sales_data = _get_daily_sales(db, days=90)

    if len(sales_data) >= 7:
        try:
            import pandas as pd
            from prophet import Prophet

            df = pd.DataFrame({
                "ds": pd.to_datetime([s["date"] for s in sales_data]),
                "y": [s["amount"] for s in sales_data],
            })
            model = Prophet(
                yearly_seasonality=False,
                weekly_seasonality=True,
                daily_seasonality=False,
                changepoint_prior_scale=0.05,
                uncertainty_samples=100,
            )
            model.fit(df)
            future = model.make_future_dataframe(periods=days)
            forecast = model.predict(future)
            future_fc = forecast.tail(days)

            predictions = [
                {
                    "date": row["ds"].date().isoformat(),
                    "predicted_amount": round(max(0, row["yhat"]), 0),
                    "confidence_low": round(max(0, row["yhat_lower"]), 0),
                    "confidence_high": round(max(0, row["yhat_upper"]), 0),
                }
                for _, row in future_fc.iterrows()
            ]
            return {"predictions": predictions, "source": "prophet"}
        except Exception:
            pass  # Prophet موجود نیست یا خطا داد → fallback

    return _moving_average_predict(sales_data, days)


def get_trend(sales_data: List[Dict]) -> str:
    """تعیین روند کلی فروش"""
    if len(sales_data) < 4:
        return "stable"
    half = len(sales_data) // 2
    first_avg = sum(s["amount"] for s in sales_data[:half]) / half
    second_avg = sum(s["amount"] for s in sales_data[half:]) / half
    if first_avg == 0:
        return "stable"
    change = (second_avg - first_avg) / first_avg
    if change > 0.05:
        return "rising"
    if change < -0.05:
        return "falling"
    return "stable"


_PERSIAN_WEEKDAYS = {
    0: "دوشنبه", 1: "سه‌شنبه", 2: "چهارشنبه",
    3: "پنجشنبه", 4: "جمعه", 5: "شنبه", 6: "یکشنبه",
}


def get_best_worst_day(sales_data: List[Dict]) -> tuple[str, str]:
    """بهترین و بدترین روز هفته از نظر فروش"""
    day_totals: Dict[int, float] = defaultdict(float)
    day_counts: Dict[int, int] = defaultdict(int)

    for s in sales_data:
        d = date.fromisoformat(s["date"])
        day_totals[d.weekday()] += s["amount"]
        day_counts[d.weekday()] += 1

    if not day_totals:
        return "—", "—"

    avgs = {k: day_totals[k] / day_counts[k] for k in day_totals}
    best = max(avgs, key=avgs.get)
    worst = min(avgs, key=avgs.get)
    return _PERSIAN_WEEKDAYS.get(best, "—"), _PERSIAN_WEEKDAYS.get(worst, "—")


def get_stock_alerts(db: Session) -> List[Dict]:
    """هشدار موجودی — کاملاً داینامیک از دیتابیس"""
    from models.product import Product
    from models.invoice import Invoice, InvoiceItem

    thirty_days_ago = datetime.utcnow() - timedelta(days=30)

    # میانگین روزانه فروش هر محصول
    sales_rows = (
        db.query(
            InvoiceItem.product_id,
            func.sum(InvoiceItem.quantity).label("total_qty"),
        )
        .join(Invoice, Invoice.id == InvoiceItem.invoice_id)
        .filter(Invoice.created_at >= thirty_days_ago)
        .filter(Invoice.status == "completed")
        .group_by(InvoiceItem.product_id)
        .all()
    )
    avg_daily: Dict[int, float] = {r.product_id: r.total_qty / 30.0 for r in sales_rows}

    products = db.query(Product).filter(Product.is_active == True).all()
    alerts = []
    for p in products:
        daily = avg_daily.get(p.id, 0)
        if daily <= 0:
            continue
        days_left = p.stock_quantity / daily
        if days_left > 7:
            continue
        urgency = "critical" if days_left <= 2 else ("warning" if days_left <= 5 else "info")
        alerts.append({
            "product_id": p.id,
            "product_name": p.name,
            "current_stock": p.stock_quantity,
            "daily_avg_sales": round(daily, 1),
            "days_remaining": round(days_left, 1),
            "urgency": urgency,
        })

    return sorted(alerts, key=lambda x: x["days_remaining"])


def get_top_products_next_week(db: Session) -> Dict:
    """پیش‌بینی پرفروش‌ترین محصولات هفته آینده"""
    from models.invoice import Invoice, InvoiceItem
    from models.product import Product

    now = datetime.utcnow()
    this_week_start = now - timedelta(days=7)
    prev_week_start = now - timedelta(days=14)

    def _week_sales(start, end):
        rows = (
            db.query(
                InvoiceItem.product_id,
                InvoiceItem.product_name,
                func.sum(InvoiceItem.quantity).label("qty"),
                func.sum(InvoiceItem.subtotal).label("revenue"),
            )
            .join(Invoice, Invoice.id == InvoiceItem.invoice_id)
            .filter(Invoice.created_at >= start)
            .filter(Invoice.created_at < end)
            .filter(Invoice.status == "completed")
            .group_by(InvoiceItem.product_id, InvoiceItem.product_name)
            .order_by(func.sum(InvoiceItem.revenue).desc())
            .limit(5)
            .all()
        )
        return {r.product_id: {"name": r.product_name, "qty": float(r.qty or 0), "revenue": float(r.revenue or 0)} for r in rows}

    current = _week_sales(this_week_start, now)
    previous = _week_sales(prev_week_start, this_week_start)

    if not current:
        return {"next_week_tops": [], "data_available": False, "message": "داده کافی برای پیش‌بینی وجود ندارد"}

    tops = []
    for pid, data in current.items():
        prev_qty = previous.get(pid, {}).get("qty", 0)
        if prev_qty > 0:
            trend_pct = (data["qty"] - prev_qty) / prev_qty * 100
        else:
            trend_pct = 100.0

        sign = "+" if trend_pct >= 0 else ""
        tops.append({
            "product_id": pid,
            "product_name": data["name"],
            "predicted_sales": round(data["qty"] * 1.05, 1),  # پیش‌بینی هفته آینده ≈ +۵٪
            "trend_percent": round(trend_pct, 1),
            "trend_label": f"{sign}{round(trend_pct,0):.0f}٪ نسبت به هفته قبل",
        })

    return {"next_week_tops": tops, "data_available": True}


def build_sales_summary(db: Session, period: str) -> Dict:
    """ساخت خلاصه فروش برای ارسال به Claude (کاملاً داینامیک)"""
    from models.invoice import Invoice, InvoiceItem

    days = 7 if period == "week" else 30
    now = datetime.utcnow()
    curr_start = now - timedelta(days=days)
    prev_start = now - timedelta(days=days * 2)

    def _period_stats(start, end):
        rows = (
            db.query(
                func.sum(Invoice.final_amount).label("total"),
                func.count(Invoice.id).label("count"),
                func.avg(Invoice.final_amount).label("avg"),
            )
            .filter(Invoice.created_at >= start)
            .filter(Invoice.created_at < end)
            .filter(Invoice.status == "completed")
            .first()
        )
        return {
            "total": float(rows.total or 0),
            "count": int(rows.count or 0),
            "avg": float(rows.avg or 0),
        }

    curr = _period_stats(curr_start, now)
    prev = _period_stats(prev_start, curr_start)

    # پرفروش‌ترین محصولات
    top_rows = (
        db.query(
            InvoiceItem.product_name,
            func.sum(InvoiceItem.quantity).label("qty"),
        )
        .join(Invoice, Invoice.id == InvoiceItem.invoice_id)
        .filter(Invoice.created_at >= curr_start)
        .filter(Invoice.status == "completed")
        .group_by(InvoiceItem.product_name)
        .order_by(func.sum(InvoiceItem.quantity).desc())
        .limit(5)
        .all()
    )
    top_products = "\n".join(f"- {r.product_name}: {int(r.qty)} عدد" for r in top_rows) or "اطلاعاتی موجود نیست"

    # محصولات رو به اتمام
    alerts = get_stock_alerts(db)[:3]
    low_stock = "\n".join(
        f"- {a['product_name']}: {a['days_remaining']} روز دیگر تمام می‌شود"
        for a in alerts
    ) or "همه محصولات موجودی کافی دارند"

    # بهترین/بدترین روز
    sales_data = _get_daily_sales(db, days=30)
    best_day, worst_day = get_best_worst_day(sales_data)

    return {
        "period_label": "۷ روز اخیر" if period == "week" else "۳۰ روز اخیر",
        "current_period_sales": f"{curr['total']:,.0f}",
        "previous_period_sales": f"{prev['total']:,.0f}",
        "invoice_count": curr["count"],
        "avg_invoice_amount": f"{curr['avg']:,.0f}",
        "top_products": top_products,
        "low_stock_products": low_stock,
        "best_day": best_day,
        "worst_day": worst_day,
    }


def get_sales_hash(db: Session) -> str:
    """هش داده‌های فروش برای تشخیص تغییر"""
    sales_data = _get_daily_sales(db, days=7)
    data_str = json.dumps(sales_data, sort_keys=True)
    return hashlib.md5(data_str.encode()).hexdigest()
